Titles and metadata

Otju opened the pull request:
Closes #46 
Added:
- Title for each page
- meta description tag for landing page and normal pages, that currently just consist of the first sentence of the page. (Could theoretically be it's own Strapi field, but I don't think it's worth it).
- Favicon using gatsby-plugin-manifest that changes color depending on whether or not the user is using dark mode.
- html lang tag that's either "fi" or "en", depending on the pages url. Might not work on the 404 page.

Otju committed:
Added favicon using gatsby-plugin-manifest

Otju committed:
Added title tags for all pages, and description tags for most pages.

Otju committed:
Added html lang tag

PurkkaKoodari commented on src/utils/firstSentenceFromContent.ts:
```diff
@@ -0,0 +1,16 @@
+import { StrapiPageContentBlock, TextBlock } from "../types/strapi"
+
+const firstSentenceFromContent = (content: StrapiPageContentBlock[]) => {
+  const textBlock = content.find(
+    (block) => block.strapi_component === "common-content.text-block"
+  ) as TextBlock | undefined
+  const firstTextBlock = textBlock?.text.data.text
+  const firstSentence = firstTextBlock
+    ?.replace(/#+.+\n/g, "") // Remove headings
+    ?.split(/[.\n]/) // Split by sentence or line break
+    .find((string) => string && string.length > 0) // Find first non-empty string
+    ?.concat(".")
+  return firstSentence
+}
```
This keeps Markdown markup, such as links, images or styling.

I think we're better off omitting `description` on content pages, or using a Markdown parser if we insist to have one autogenerated.

Also, it might be better to get a full paragraph - whatever is displaying the `description` is usually smart enough to clip it itself.

PurkkaKoodari commented on tsconfig.eslint.json:
```diff
@@ -7,6 +7,7 @@
     "gatsby-browser.js",
     "gatsby-config.js",
     "gatsby-node.js",
+    "gatsby-ssr.js",
```
This should probably be turned into `gatsby-*.js` if eslint supports it.

PurkkaKoodari commented on gatsby-ssr.js:
```diff
@@ -0,0 +1,8 @@
+  /* pathname always return "/" in development, but works with gatsby build.
+  https://github.com/gatsbyjs/gatsby/issues/4350
+  */
+  const lang = pathname.startsWith("/en/") || pathname === "/en" ? "en" : "fi"
```
At this point we should move the language-from-pathname extraction logic into `paths.ts` as it's starting to be used in quite a few places.

PurkkaKoodari commented on src/components/Meta.tsx:
```diff
@@ -0,0 +1,15 @@
+const Meta: React.FC<Props> = ({ title, description }) => {
+  return (
+    <>
+      <title>Tietokilta{title && ` - ${title}`}</title>
```
This should be the other way around so that the title _starts_ with the most accurate part. Especially useful for finding stuff when your tabs look like this.
![image](https://user-images.githubusercontent.com/8439661/209178306-24a05cc5-8a0e-49a4-858e-c181067cbdcd.png)

PurkkaKoodari commented on src/pages/404.tsx:
```diff
@@ -11,4 +13,11 @@ const NotFoundPage: React.FC<PageProps> = (props: PageProps) => {
+  const { pathname } = useLocation()
+  const locale = pathname.startsWith("/en/") ? "en" : "fi"
+  return (
+    <Meta title={locale === "fi" ? "Sivua ei löytynyt" : "Page not found"} />
```
This might benefit from a `<meta name="robots" content="noindex">` option. 404 pages are probably fine but we might see an use for them with e.g. unlisted pages as well.

PurkkaKoodari replied:
This might benefit from a `<link rel="noindex">` option. 404 pages are probably fine without one in most search engines, but we might see an use for them with e.g. unlisted pages as well.

PurkkaKoodari commented on src/templates/landingPage.tsx:
```diff
@@ -119,4 +121,19 @@ export const pageQuery = graphql`
+  const description = firstSentenceFromContent(content)
+  return (
+    <Meta
+      title={locale === "fi" ? "Etusivu" : "Front Page"}
```
Just the default title of "Tietokilta" is much better than "Front Page".

Or, we might want to use e.g. "Tietokilta &ndash; Computer Science Guild of Aalto University" on the landing page; that'll require some more changes to `Meta`.
```suggestion
```

PurkkaKoodari reviewed (changes requested):
Good start.

I think that we're better off adding a `description` field to Strapi or omitting it on content pages. At least on the landing page, which ought to be our most visible page on search engines, that'll probably be useful. (At least the current solution here is not sufficient yet.)

PurkkaKoodari commented:
Another potentially useful thing: `<link rel="alternate" hreflang="..." href="...">`

https://developers.google.com/search/docs/specialty/international/localized-versions#html

kahlstrm committed:
fixes

PurkkaKoodari committed:
Add meta noindex to 404 page

PurkkaKoodari committed:
Shorter title on landing page

PurkkaKoodari committed:
Refactor locale-from-path code into a function

PurkkaKoodari committed:
Remove broken firstSentenceFromContent for now

PurkkaKoodari committed:
Add <link rel="alternate"> for other locale

PurkkaKoodari committed:
Fix errors from gatsby inference and populate=deep

populate=deep is incompatible gatsby-source-strapi. Also,
gatsby-source-strapi is stupid.

- When loading Strapi Components of "regular" objects,
  gatsby-source-strapi converts them to Gatsby Nodes recursively using
  createNode.
- This also renames e.g. the `content` field of `Page` to
  `content___NODE` to mark it as a Gatsby relation.
- When loading relation fields, this doesn't happen. Instead,
  gatsby-source-strapi just createNode()s the related object as-is,
  but only if it doesn't exist.
- If `LandingPage` is populated deeply, the `buttonLink` relation
  contains a `Page`, which further has the `content` field populated.
- If `LandingPage` then ends up loaded before `Page`, that related
  `Page` is createNode()'d with its `content` as JSON (not converted to
  Nodes).
- After this, when `Page`s are loaded, they're created normally as
  STRAPI_PAGE Nodes with a `content___NODE` field, which would also
  appear in GraphQL as `content`. (This shows up as a console warning.)
- This `content` mismatch messes up Gatsby's node creation or inference
  or something, and the `buttonLink`ed STRAPI_PAGE ends up with null
  content. I spent a few hours tracing through this spaghetti code.

This commit explicitly specifies the population to avoid the above
clusterfuck. It still populates `buttonLink`, but only with `id`.

This *seems* to work fine for now. I'm not sure of the implications of
the junk Node created by the relation: I'm not sure if it's overwritten
or merged with the real STRAPI_PAGE created later. This has implications
later for getting `buttonLink` to work at all in a reasonable way.

In general, Strapi relations (especially as handled by
gatsby-source-strapi) don't seem to work well with Gatsby's GraphQL.

PurkkaKoodari commented:
This'll leave the event pages without dynamic titles. That'll be a new issue most likely; it seems to require a traditional `react-use` `useTitle` or `react-helmet`  based on https://github.com/gatsbyjs/gatsby/discussions/36551

PurkkaKoodari reviewed (approved):
My previous comments have been addressed - maybe someone could look through my changes still.

kahlstrm reviewed (approved):
Looks good apart, we should probably disable eslint in the pagecontext

kahlstrm committed:
eslint-disable PageContext null-assertion
